crates/holochain_cascade/src/tests.rs

```diff
@@ -0,0 +1,289 @@
+
+    // Set up the mock network to return NoPeersForLocation error
+    let action_hash_2 = action_hash.clone();
+    let action_hash_3 = action_hash.clone();
+    network
+        .expect_get_by_op_type()
+        .return_once(move |action_hash_param, op_type_param| {
+            assert_eq!(
+                action_hash_param, action_hash_2,
+                "get_by_op_type called with unexpected action hash"
+            );
+            assert_eq!(
+                op_type_param, op_type,
+                "get_by_op_type called with unexpected op type"
+            );
+            Err(HolochainP2pError::NoPeersForLocation(
+                "".to_string(),
+                action_hash_3.get_loc(),
+            ))
+        });
+    let network = Arc::new(network);
+    let cascade = CascadeImpl::empty()
+        .with_cache(cache.to_db())
+        .with_network(network.clone(), test_cache_db().to_db());
+
+    // Call the method
+    let result = cascade
+        .fetch_op_by_type(action_hash, op_type)
+        .await
+        .unwrap();
+
+    // Verify that None is returned when there are no peers
+    assert!(result.is_none());
+}
+
+#[tokio::test(flavor = "multi_thread")]
+async fn network_error_returns_error() {
+    let cache = test_cache_db();
+    let mut network = MockHolochainP2pDnaT::new();
+
+    let action_hash = fixt!(ActionHash);
+    let op_type = ChainOpType::StoreRecord;
+
+    // Set up the mock network to return a generic network error
+    let action_hash_2 = action_hash.clone();
+    network
+        .expect_get_by_op_type()
+        .return_once(move |action_hash_param, op_type_param| {
+            assert_eq!(
+                action_hash_param, action_hash_2,
+                "get_by_op_type called with unexpected action hash"
+            );
+            assert_eq!(
+                op_type_param, op_type,
+                "get_by_op_type called with unexpected op type"
+            );
+            Err(HolochainP2pError::invalid_p2p_message("ohono".to_string()))
+        });
+    let network = Arc::new(network);
+    let cascade = CascadeImpl::empty()
+        .with_cache(cache.to_db())
+        .with_network(network.clone(), test_cache_db().to_db());
+
+    // Call the method and expect an error
+    let result = cascade.fetch_op_by_type(action_hash, op_type).await;
+
+    // Verify that an error is returned
+    assert!(result.is_err());
+}
+
+#[tokio::test(flavor = "multi_thread")]
+async fn no_network_returns_none() {
+    let cache = test_cache_db();
+    let cascade = CascadeImpl::empty().with_cache(cache.to_db());
+
+    let action_hash = fixt!(ActionHash);
+    let op_type = ChainOpType::StoreRecord;
+
+    // Call without a network
+    let result = cascade.fetch_op_by_type(action_hash, op_type).await;
+
+    // Verify that None is returned if no network is present.
+    assert!(matches!(result, Ok(None)));
+}
+
+#[tokio::test(flavor = "multi_thread")]
+async fn fetched_op_is_added_to_cache() {
```

Comment:
Test that fetched ops are added to the cache, which is the main side-effect of the call.